extensions/guacamole-auth-totp/src/main/java/org/apache/guacamole/auth/totp/user/CodeUsageTrackingService.java

```diff
@@ -0,0 +1,264 @@
+         * used by the given user.
+         *
+         * @param username
+         *     The username of the user which previously used the given code.
+         *
+         * @param code
+         *     The valid code given by the user.
+         */
+        public UsedCode(String username, String code) {
+            this.username = username;
+            this.code = code;
+        }
+
+        /**
+         * Returns the username of the user which previously used the code
+         * associated with this UsedCode.
+         *
+         * @return
+         *     The username of the user which previously used this code.
+         */
+        public String getUsername() {
+            return username;
+        }
+
+        /**
+         * Returns the valid code given by the user when this UsedCode was
+         * created.
+         *
+         * @return
+         *     The valid code given by the user.
+         */
+        public String getCode() {
+            return code;
+        }
+
+        @Override
+        public int hashCode() {
+            int hash = 7;
+            hash = 79 * hash + this.username.hashCode();
+            hash = 79 * hash + this.code.hashCode();
+            return hash;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+
+            if (this == obj)
+                return true;
+
+            if (obj == null)
+                return false;
+
+            if (getClass() != obj.getClass())
+                return false;
+
+            final UsedCode other = (UsedCode) obj;
+            return username.equals(other.username) && code.equals(other.code);
+
+        }
+
+    }
+
+    /**
+     * Attempts to mark the given code as used. The code MUST have already been
+     * validated against the user's secret key, as this function only verifies
+     * whether the code has been previously used, not whether it is actually
+     * valid. If the code has not previously been used, the code is stored as
+     * having been used by the given user at the current time.
+     *
+     * @param username
+     *     The username of the user who has attempted to use the given valid
+     *     code.
+     *
+     * @param code
+     *     The otherwise-valid code given by the user.
+     *
+     * @return
+     *     true if the code has not previously been used by the given user and
+     *     has now been marked as previously used, false otherwise.
+     *
+     * @throws GuacamoleException
+     *     If configuration information necessary to determine the length of
+     *     time a code should be marked as invalid cannot be read from
+     *     guacamole.properties.
+     */
+    public boolean useCode(String username, String code)
+            throws GuacamoleException {
+
+        // Repeatedly attempt to use the given code until an explicit success
+        // or failure has occurred
+        UsedCode usedCode = new UsedCode(username, code);
+        for (;;) {
```

Comment:
Any reason this is a `for` loop and not `while (true)`?

Reply:
Only that it's what I'm used to. If you think `while (true)` is more clear, I'm happy to switch things.

Reply:
Nah, doesn't matter much to me, just curious.